File: s_box_generation.py
from PRNG import chaotic_PRNG
import numpy as np

# Trích chữ số thập phân
def extract_digits(x, pos=[3,4,5]):
    s = f"{x:.10f}".replace("0.","")  # lấy 10 chữ số thập phân
    digits = "".join(s[p-1] for p in pos)  # pos tính từ 1
    return int(digits)

# Hàm nghịch đảo trong GF(2^8) với đa thức AES (0x11B)
def gf_inverse(x):
    if x == 0:
        return 0
    a, m = x, 0x11B
    u, v = 1, 0
    while a > 1:
        shift = a.bit_length() - m.bit_length()
        if shift >= 0:
            a ^= m << shift
            u ^= v << shift
        else:
            a, m = m, a
            u, v = v, u
    return u & 0xFF

# Tạo S-Box (256 phần tử)
def build_sbox(seq, needed=256):
    sbox = []
    seen = set()
    for (x1, x2, x3) in seq:
        D1 = extract_digits(x1)
        D2 = extract_digits(x2)
        D3 = extract_digits(x3)
        M1, M2, M3 = D1 % 256, D2 % 256, D3 % 256
        V = gf_inverse(M1 ^ M2 ^ M3)
        if V not in seen:
            seen.add(V)
            sbox.append(V)
        if len(sbox) == needed:
            break
    return sbox

# Lưu S-Box ra CSV
def save_sbox_csv(sbox, path="sbox_from_algorithm.csv"):
    mat = np.array(sbox).reshape(16,16)
    with open(path, "w", newline="") as f:
        writer = csv.writer(f)
        for row in mat:
            writer.writerow(row)
    return path

# In S-Box 16x16
def print_sbox(sbox):
    print("S-BOX (16x16):")
    for i in range(16):
        row = []
        for j in range(16):
            idx = 16*i + j
            if idx < len(sbox):
                row.append(f"{sbox[idx]:02X}")
            else:
                row.append("..")  
        print(" ".join(row))

# Sinh key từ S-Box
def sbox_to_key(sbox, key_size=16):
    key_bytes = bytes(sbox[:key_size])
    return key_bytes

File: C/s_box_generation.c
#include "PRNG.c"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

int extract_digits(double x) {
	char buf[64];
	snprintf(buf, sizeof(buf), "%.10f", x);
	char *p = strchr(buf, '.');
	if (!p) return 0;
	p++;

	char digits[4];
	for (int i = 0; i < 3; i++) {
		if ((p + 2 + i) < buf + strlen(buf))
			digits[i] = p[2 + i]; 
		else
		    	digits[i] = '0';
	}
	digits[3] = '\0';
	return atoi(digits); 
}

void build_sbox(/* Chaotic system parameterss */ 	double A[3][3], int sample, int N, double epsilon, 
		/* Initial vector of state variables*/ 	double x0[3], 
		/* A S-box matrix of size nxm */ 	unsigned char sbox[256]){

	int j = 0; 	

    	//double (*seq)[3] = malloc(sample * sizeof(*seq));
	double (*seq)[3] = (double (*)[3]) malloc(sample * sizeof(*seq));

	chaotic_PRNG(A, sample, N, epsilon, x0, seq);

	int k = 0;
	int used[256];
 	memset(used, 0, sizeof(used));

	while(j < 256 && k < sample){
		int D1 = extract_digits(seq[k][0]);
		int D2 = extract_digits(seq[k][1]);
		int D3 = extract_digits(seq[k][2]);
		k++;

		int M1 = D1 % 256;
		int M2 = D2 % 256;
		int M3 = D3 % 256;

		int V = (M1 ^ M2 ^ M3);

		if(!used[V]){
			used[V] = 1;
			sbox[j++] = (unsigned char)V;
		}
	}

	free(seq);
}

// Lưu S-Box ra CSV
void save_sbox_csv(int *sbox, int len, const char *path) {
    FILE *f = fopen(path, "w");
    if (!f) {
        printf("Không mở được file %s\n", path);
        return;
    }
    for (int i = 0; i < 16; i++) {
        for (int j = 0; j < 16; j++) {
            int idx = i*16 + j;
            if (idx < len) fprintf(f, "%d", sbox[idx]);
            else fprintf(f, "0");
            if (j < 15) fprintf(f, ",");
        }
        fprintf(f, "\n");
    }
    fclose(f);
}

// In S-Box dạng 16x16 hex
void print_sbox(const unsigned char *sbox, int len) {
    printf("S-BOX (16x16):\n");
    for (int i = 0; i < 16; i++) {
        for (int j = 0; j < 16; j++) {
            int idx = i * 16 + j;
            if (idx < len)
                printf("%02X ", sbox[idx]);
            else
                printf(".. ");
        }
        printf("\n");
    }
}

// Sinh key từ S-Box (lấy 16 byte đầu)
void sbox_to_key(int *sbox, unsigned char *key, int key_size) {
    for (int i = 0; i < key_size; i++) {
        key[i] = (unsigned char)sbox[i];
    }
}

int main() {
    double A[3][3] = {
        {0.0, 1.0/2, 1.0/20},
        {1.0/3, 0.0, 1.0/3},
        {1.0/20, 1.0/2, 0.0}
    };
    int sample = 2000;
    int N = 100;
    double seq[sample][3];
    double x0[3] = {0.1, 0.01, 0.0};

    unsigned char sbox[256];
    build_sbox(A, sample, N, 0.05, x0, sbox);

    print_sbox(sbox, 256);


    return 0;
}

